File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        maxDepth_ = 0

        if not root:
            return maxDepth_

        stack = [(root, 1)]
        maxDepth_ = 1
        seen = set()
        
        while stack:
            parent, depth  = stack[-1]

            if parent.left and parent not in seen:
                stack.append((parent.left, depth + 1))

                maxDepth_ = max(maxDepth_, depth + 1)
                seen.add(parent)
            
            else:
                parent, depth = stack.pop()
                if parent.right:
                    stack.append((parent.right, depth + 1))
                    maxDepth_ = max(maxDepth_, depth + 1)
        
        return maxDepth_

File: Data Structures & Algorithms/valid-binary-search-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    boolean isValid = true;
    long prev = Long.MIN_VALUE;

    public void dfs(TreeNode root) {
        if (root == null) {
            return;
        }
        dfs(root.left);

        if (!(prev < root.val)) {
            isValid = false;
        }
        prev = root.val;
        dfs(root.right);
    }
    public boolean isValidBST(TreeNode root) {

        dfs(root);
        return isValid;
        
    }
}
